weed/server/volume_server_handlers.go

```diff
@@ -36,55 +36,76 @@ func (vs *VolumeServer) privateStoreHandler(w http.ResponseWriter, r *http.Reque
 		w.Header().Set("Access-Control-Allow-Origin", "*")
 		w.Header().Set("Access-Control-Allow-Credentials", "true")
 	}
+
+	// track request time
 	stats.VolumeServerRequestCounter.WithLabelValues(r.Method).Inc()
 	start := time.Now()
 	defer func(start time.Time) {
 		stats.VolumeServerRequestHistogram.WithLabelValues(r.Method).Observe(time.Since(start).Seconds())
 	}(start)
+
 	switch r.Method {
 	case "GET", "HEAD":
 		stats.ReadRequest()
-		vs.inFlightDownloadDataLimitCond.L.Lock()
-		for vs.concurrentDownloadLimit != 0 && atomic.LoadInt64(&vs.inFlightDownloadDataSize) > vs.concurrentDownloadLimit {
-			select {
-			case <-r.Context().Done():
-				glog.V(4).Infof("request cancelled from %s: %v", r.RemoteAddr, r.Context().Err())
-				return
-			default:
-				glog.V(4).Infof("wait because inflight download data %d > %d", vs.inFlightDownloadDataSize, vs.concurrentDownloadLimit)
-				vs.inFlightDownloadDataLimitCond.Wait()
+
+		inFlightDownloadSize := atomic.LoadInt64(&vs.inFlightDownloadDataSize)
+		if vs.concurrentDownloadLimit != 0 && inFlightDownloadSize > vs.concurrentDownloadLimit {
+			for {
+				glog.V(4).Infof("wait because inflight download data %d > %d", inFlightDownloadSize, vs.concurrentDownloadLimit)
+				select {
+				case <-r.Context().Done():
+					glog.V(4).Infof("request cancelled from %s: %v", r.RemoteAddr, r.Context().Err())
+					go func() {
+						<-vs.inFlightDownloadWaitCh
```

Comment:
I could not follow the logic easily here. Why an extra goroutine is needed, which is much heavier than atomic operations.

Maybe do some benchmark to see whether this is faster.

Reply:
This is needed in case the request was canceled, to make sure the incoming signal is drained from the channel. But if you don't like this I can make it `Cond` again. I just thought channels are more idiomatic.